File: tianshu/core/audit.py
from __future__ import annotations

import json
import threading
import time
from datetime import UTC, datetime

from tianshu.config import WORKSPACE_DIR

AUDIT_DIR = WORKSPACE_DIR / ".ts-audit"
_lock = threading.Lock()


def audit(event: str, detail: str, actor: str = "system") -> None:
    try:
        AUDIT_DIR.mkdir(parents=True, exist_ok=True)
        line = json.dumps(
            {"ts": datetime.now(UTC).isoformat(timespec="seconds"), "event": event, "actor": actor, "detail": detail},
            ensure_ascii=False,
        )
        with _lock, open(AUDIT_DIR / f"audit-{time.strftime('%Y%m%d')}.log", "a", encoding="utf-8") as f:
            f.write(line + "\n")
    except OSError:
        pass

File: tianshu-ts/src/core/audit.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { WORKSPACE_DIR } from "../config.js";

export const AUDIT_DIR = path.join(WORKSPACE_DIR, ".ts-audit");

export function audit(event: string, detail: string, actor = "system"): void {
  try {
    fs.mkdirSync(AUDIT_DIR, { recursive: true });
    const line = JSON.stringify({
      ts: new Date().toISOString(),
      event,
      actor,
      detail,
    });
    const day = new Date().toISOString().slice(0, 10).replace(/-/g, "");
    fs.appendFileSync(path.join(AUDIT_DIR, `audit-${day}.log`), line + "\n", "utf-8");
  } catch {
    // 审计失败不阻断主流程
  }
}